### DISCUSSION TITLE
Solver Equation Sequence - Example: Heat pump

### DISCUSSION THREAD

**Discussion by RouvenZ** [OpenModelica/OpenModelica#10830] [Q&A] [answered]
Hello everyone,

is there a possibility to see in which sequence the solver (in my case the default [DASSL](https://www.osti.gov/servlets/purl/5882821)) is evaluating the equations in the code generated from the compiler and with which values they are evaluated? I see equation indexes in the Equations Browser and find a function `functionODE_system0` within the generated C-code which seems to call the functions of the respective equations in order of their assigned indices. However, I was not able to see results of the individual function calls.

My question arises from the [heat pump scroll compressor validation model ](https://github.com/lbl-srg/modelica-buildings/blob/115ea729ea19eb45ff817b40e2276ed8cd49e1b4/Buildings/Fluid/HeatPumps/Validation/ScrollWaterToWater_Dynamic.mo) in the [buildings library](https://github.com/lbl-srg/modelica-buildings). I switched one heat pump to`energyDynamics = ...fixedInitial`.
As expected, a torn equation system arises comprising evaporator, compressor and condenser. Due to the non-zero evaporator/condenser volume, the volume temperature $T$ is forming a differential equation. This equation depends on the heat flow rate $Q$ being exchanged, which is part of the torn equation system. On the other hand, the volume temperature $T$ is necessary to solve the torn equation system. 
However, the differential equation is not considered to be part of the torn system by OpenModelica and evaluated separately. I assume this to be because DASSL handles differential equations differently and propagates solutions via the backward differentiation formula (BDF) along timesteps, so that either of these equation "nodes" is evaluated with values extrapolated from prior timesteps (or starting values). However, to understand the underlying processes in detail, I would like to know what values for $T$ are used to solve the torn system (and obtain $Q$) and what values for $Q$ are used to solve the differential equation (and obtain $T$). As (according to the `### STATISTICS ###`) my `functionODE` is called more often than there are steps taken, I assume that process to repeat at some timesteps until convergence is reached. 

Thank you a lot for your reply and have a nice day!

Cheers
Rouven

**Comment by casella** [comment, marked as answer]
You can see the sequence of evaluations using the [Equation-based Debugger](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/debugger.html#the-equation-based-debugger). This is actually not DASSL running, it's the OpenModelica-generated code that computes all the derivatives that are then used by DASSL to advance to the next state.

Currently the equation-based debugger only shows static information. It is planned to add run-time information (e.g., variable values) in the next 12 months.

**Reply by RouvenZ** [reply]
Thank you a lot for your reply! 
I have indeed consulted the [Equation-based Debugger](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/debugger.html#the-equation-based-debugger) already, but was not sure whether the "Index" there in the [lower left part in the "Equations Browser"](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/_images/omedit-transformationsbrowser.png) indeed reflects the sequence of evaluations. If I got your answer right, this is indeed the case, or am I mistaken? 

Thank you moreover for showing a path towards run-time information, that would be highly appreciated!

Have a nice week!

**Comment by casella** [comment]
The index is just a label, but it also corresponds to the computational sequence _among the same type_, e.g. initial, regular, initial-lambda0, etc.

**Comment by casella** [comment]
It's also used for naming the C functions that actually carry out the computations.
